src/Nethermind/Nethermind.Blockchain/BlockTree.cs

```diff
@@ -1268,32 +1268,60 @@ public Block FindBlock(Keccak blockHash, BlockTreeLookupOptions options)
 
         private void SetTotalDifficulty(BlockHeader header)
         {
+            BlockHeader GetParentHeader(BlockHeader current) => 
+                // TotalDifficultyNotNeeded is by design here,
+                // if it was absent this would result in recursion, as if parent doesn't already have total difficulty 
+                // then it would call back to SetTotalDifficulty for it
+                // This was original code but it could result in stack overflow
+                this.FindParentHeader(current, BlockTreeLookupOptions.TotalDifficultyNotNeeded) 
+                ?? throw new InvalidOperationException($"An orphaned block on the chain {current}");
+
+            void SetTotalDifficultyDeep(BlockHeader current)
+            {
+                Stack<BlockHeader> stack = new Stack<BlockHeader>();
+                while (current.TotalDifficulty == null)
+                {
+                    (BlockInfo blockInfo, ChainLevelInfo level) = LoadInfo(current.Number, current.Hash, true);
+                    if (level == null || blockInfo == null)
+                    {
+                        stack.Push(current);
+                        if (_logger.IsTrace) _logger.Trace($"Calculating total difficulty for {current.ToString(BlockHeader.Format.Short)}");
+                        current = GetParentHeader(current);
+                    }
+                    else
+                    {
+                        current.TotalDifficulty = blockInfo.TotalDifficulty;
+                    }
+                }
+
+                while (stack.TryPop(out BlockHeader child))
+                {
+                    child.TotalDifficulty = current.TotalDifficulty + child.Difficulty;
+                    BlockInfo blockInfo = new BlockInfo(child.Hash, child.TotalDifficulty.Value);
```

Comment:
is BlockInfo ever cached? shall we operate on the original object?

Reply:
There was no original object, block info was missing